server/sonar-server/src/test/java/org/sonar/server/computation/debt/DebtModelHolderImplTest.java

```diff
@@ -0,0 +1,122 @@
+
+    assertThat(sut.rootCharacteristics()).hasSize(2);
+    assertThat(sut.subCharacteristicsByRootKey("PORTABILITY")).hasSize(2);
+  }
+
+  @Test
+  public void add_characteristics_fail_with_a_NPE_if_root_characteristic_is_null() throws Exception {
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage("rootCharacteristic cannot be null");
+
+    sut.addCharacteristics(null, Collections.singletonList(new CharacteristicImpl(2, "COMPILER_RELATED_PORTABILITY")));
+  }
+
+  @Test
+  public void add_characteristics_fail_with_a_NPE_if_sub_characteristics_are_null() throws Exception {
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage("subCharacteristics cannot be null");
+
+    sut.addCharacteristics(new CharacteristicImpl(1, "PORTABILITY"), null);
+  }
+
+  @Test
+  public void add_characteristics_fail_with_a_ISE_if_sub_characteristics_are_empty() throws Exception {
+    thrown.expect(IllegalStateException.class);
+    thrown.expectMessage("subCharacteristics cannot be empty");
+
+    sut.addCharacteristics(new CharacteristicImpl(1, "PORTABILITY"), Collections.<Characteristic>emptyList());
+  }
+
+  @Test
+  public void get_characteristic_by_key() throws Exception {
+    sut.addCharacteristics(new CharacteristicImpl(1, "PORTABILITY"), Collections.singletonList(new CharacteristicImpl(2, "COMPILER_RELATED_PORTABILITY")));
+
+    assertThat(sut.characteristicByKey("PORTABILITY")).isNotNull();
+    assertThat(sut.characteristicByKey("COMPILER_RELATED_PORTABILITY")).isNotNull();
+    assertThat(sut.characteristicByKey("UNKNOWN")).isNull();
+  }
+
+  @Test
+  public void get_characteristic_by_key_throws_a_ISE_when_not_initialized() throws Exception {
+    thrown.expect(IllegalStateException.class);
+    thrown.expectMessage("Characteristics have not been initialized yet");
+
+    sut.characteristicByKey("PORTABILITY");
+  }
+
+  @Test
+  public void get_root_characteristics() throws Exception {
+    sut.addCharacteristics(new CharacteristicImpl(1, "PORTABILITY"),
+      Arrays.asList(new CharacteristicImpl(2, "COMPILER_RELATED_PORTABILITY"), new CharacteristicImpl(4, "READABILITY")));
+    sut.addCharacteristics(new CharacteristicImpl(3, "MAINTAINABILITY"), Collections.singletonList(new CharacteristicImpl(4, "READABILITY")));
+
+    assertThat(sut.rootCharacteristics()).hasSize(2);
+  }
+
+  @Test
+  public void get_root_characteristics_throws_a_ISE_when_not_initialized() throws Exception {
+    thrown.expect(IllegalStateException.class);
+    thrown.expectMessage("Characteristics have not been initialized yet");
+
+    sut.rootCharacteristics();
+  }
+
+  @Test
+  public void get_sub_characteristics_by_root_key() throws Exception {
+    sut.addCharacteristics(new CharacteristicImpl(1, "PORTABILITY"),
+      Arrays.asList(new CharacteristicImpl(2, "COMPILER_RELATED_PORTABILITY"), new CharacteristicImpl(4, "READABILITY")));
```

Comment:
you could use constants for `new CharacteristicImpl(1, "PORTABILITY")`, `new CharacteristicImpl(2, "COMPILER_RELATED_PORTABILITY")`, etc...